rfcs/0120-artifact-service.md

```diff
@@ -0,0 +1,393 @@
+                                            parts: [
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }, {
+                                              sha256: 'abcd1234',
+                                              size: 1024,
+                                            }],
+                                          }
+objsvc --> s3 POST /object-abcd123?uploads
+objsvc <-- s3 200 UploadId=u-123
+uploader <-- objsvc 200 {requests: [{
+                          url: 'http://s3.com/object-abcd123?partNumber=1&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }, {
+                          url: 'http://s3.com/object-abcd123?partNumber=2&uploadId=u-123',
+                          headers: {authorization: 'signature-request-1'},
+                          method: 'PUT'
+                        }
+                       ]}
+uploader --> s3 PUT /object-abcd123?partNumber=1&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 123}
+uploader --> s3 PUT /object-abcd123?partNumber=2&uploadId=u-123'
+uploader <-- s3 200 ETag {etag: 456}
+uploader --> objsvc PATCH /objects/abcd123 {etags: [123, 456]}
+objsvc --> s3 POST /object-abcd123?uploadId=u-123 {etags: [123, 456]}
+objsvc <-- s3 200 ETag {etag: 123456}
+uploader <-- objsvc 200
+```
+#### Part Size for Multipart Uploads
+Many object storage systems offer the option to perform multipart uploads.
+This is done to enable parallel uploads as well as enabling files which are
+larger than any single HTTP request ought to be.  For example, Amazon S3 limits
+object creation requests to be no greater than 5GB.  In order to upload 10GB,
+you must use a multipart upload.
+
+One source of complexity here is that the uploader needs to know what part size
+they should be using.  We will document that the suggested part size is 64MB.
+This will be a part size which we will do our best to support in all cases.  If
+a backing storage does not support 64MB, we will investigate 
+
+In the case that a backing service does not support using 64MB parts, or the
+user provides a custom part size which is not supported by the backing service,
+we will respond to the `PUT` request with a 400-series error along with a JSON
+error message that provides information on part sizes which will work.  The
+uploader should store a copy of this information to inform future uploads.
+
+### Deletion
+
+```
+DELETE /objects/:name
+```
+
+This service has at least one copy of each stored file.  Any copies stored in
+addition to this original and authoritative copy are considered to be cached
+copies.
+
+This endpoint will delete the original copy of the file as well as
+initiating the deletion of any cached copies.  On success, it will return a 200
+to reflect the success of deleting the original copy.
```

Comment:
https://tools.ietf.org/html/rfc7231#section-4.3.5 governs behavior of the `DELETE` HTTP method and this spec seems to violate it by issuing a `200` instead of a `204` since no response is sent.

This is another case where use of advanced HTTP features may be more trouble than it's worth. I recommend sticking to `GET` and `POST` unless you want to find yourself dealing with clients insisting on spec conformance.

Reply:
I've modified "200" to a  "200 series response", but you're write that 204 means an empty body but successful response :)